sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala

```diff
@@ -399,7 +400,7 @@ class InMemoryCatalog(
     // TODO: we should follow hive to roll back if one partition path failed to delete, and support
     // partial partition spec.
     partSpecs.foreach { p =>
-      if (existingParts.contains(p) && shouldRemovePartitionLocation) {
+      if (existingParts.contains(p) && shouldRemovePartitionLocation && !retainData) {
```

Comment:
shall we put the `!retainData` in the definition of `shouldRemovePartitionLocation`?